homeassistant/components/rainbird/__init__.py

```diff
@@ -61,47 +56,104 @@
-                for controller_config in config[DOMAIN]
-            ]
+    if DOMAIN not in config:
+        return True
+
+    for controller_config in config[DOMAIN]:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=controller_config,
+            )
         )
+
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2023.4.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
     )
 
+    return True
 
-async def _setup_controller(hass, controller_config, config):
-    """Set up a controller."""
-    server = controller_config[CONF_HOST]
-    password = controller_config[CONF_PASSWORD]
-    client = AsyncRainbirdClient(async_get_clientsession(hass), server, password)
-    controller = AsyncRainbirdController(client)
-    try:
-        await controller.get_serial_number()
-    except RainbirdApiException as exc:
-        _LOGGER.error("Unable to setup controller: %s", exc)
-        return False
 
-    rain_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_sensor_state)
-    delay_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_delay)
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the config entry for Rain Bird."""
 
-    for platform in PLATFORMS:
-        hass.async_create_task(
-            discovery.async_load_platform(
-                hass,
-                platform,
-                DOMAIN,
-                {
-                    RAINBIRD_CONTROLLER: controller,
-                    SENSOR_TYPE_RAINSENSOR: rain_coordinator,
-                    SENSOR_TYPE_RAINDELAY: delay_coordinator,
-                    **controller_config,
-                },
-                config,
+    hass.data.setdefault(DOMAIN, {})
+
+    controller = AsyncRainbirdController(
+        AsyncRainbirdClient(
+            async_get_clientsession(hass),
+            entry.data[CONF_HOST],
+            entry.data[CONF_PASSWORD],
+        )
+    )
+    coordinator = RainbirdUpdateCoordinator(
+        hass,
+        name=entry.title,
+        controller=controller,
+        serial_number=entry.data[CONF_SERIAL_NUMBER],
+    )
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    async def set_rain_delay(call: ServiceCall) -> None:
+        """Service call to delay automatic irrigigation."""
+        duration = call.data[ATTR_DURATION]
+        entry_ids = await async_extract_config_entry_ids(hass, call)
+        if not entry_ids:
+            raise HomeAssistantError("Device id did not match any devices")
+        await asyncio.gather(
```

Comment:
This may not be possible here, just adding a note:  A device could be connected to more than one config entry and then we'd need to check that the entry_id is from our integration domain.

Reply:
So this needs to look up all the config entries then limit to a set of config entries for this domain?  

Seems like perhaps this logic: https://github.com/home-assistant/core/blob/cb13418babd21a1e9584978b0c523f1b1e4e1cb0/homeassistant/components/fritz/services.py#L84 could be shared as well as this seems like a common operation.